### ISSUE TITLE
General Discussion

### ISSUE THREAD

**Issue by luboskmetko** [xfiveco/dev-tools#2]
A general thread for quick discussions, brainstorming ideas, etc.

**Comment by danekszy** [comment]
So the whole point of this repo is to give voice to developers and see what we're struggling with, what would help us and throw our ideas around.
Since idea for this repo was established out of our discontent with generator, I propose we put it at the center, and think what could be done better and what would help us work more efficiently...

To kick off the discussion here are some of my ideas (inc. crazy ones :smile:):
- Collect some useful SCSS mixins (many get reused from project to project & it's easier for us to remove them, rather then add, eg. svg fallbacks, retina mixins etc.)
- Rewrite generator to some new tech ([Gulp?](https://viget.com/extend/gulp-browserify-starter-faq) broccoli? sth else?)
- Individual task toggling
- Add Wordpress proj. scaffolding, with WPized perhaps? (so we don't have to download WP each time & configure it)
- Use timber templating for Wordpress projects
- Tasks for testing
- Task for svg spriting
- Visual regression
- SCSS linting (scss-lint)
- Code style checking (JSCS)
- JS linting (jshint, eslint)
- Script bundling (browserify? webpack?)
- ES6/ES7 support (babel)
- Front end templating (Jade?)
- Page speed evaluation
- Accessability evaluation (a11y)
- HTML validation
- Webfont generation
- Inlining critical path CSS
- Detect unused CSS (uncss)

Let me know what you think and don't hesitate to discuss or to propose your own.
If some ideas will get traction we'll move the discussion to separate issue.
Let the brainstorm begin! :smile:

**Comment by adamsfather** [comment]
Very good ideas Dan! :+1: 

in addition to scss reusable mixins I believe there's potential of collecting reusable html snippets and javascript interactions which are frequently used from project to project. We don't need to recreate Bootstrap or ZURB foundation, however having own standardized and fine tuned library of responsive FE elements, would save work time and avoid repetitive coding tasks, I believe. (maybe I'm totally wrong here ;P)

**Comment by danekszy** [comment]
@adamsfather Sure, like:
http://pea.rs/navigation/horizontal-tabs
http://refills.bourbon.io/components/
http://responsiveemailpatterns.com/
We could get the essentials and make them easy to import (eg. uncomment the file import etc.)
But we have to remember that the key here is to speed us up, and not bloat with unused stuff... :wink:

**Comment by adamsfather** [comment]
@danekszy you know I had an idea some time ago when generating project in generator, to have such possibility to choose which elements it will generate based on selection which corresponds with designs e.g. image slider, accordion, header nav, mobile menu, parallax-like div, page scroll, news box, service box, content box with image, etc... 

As developer you can easily recognise what elements you'll use in your code by reviewing the design. Example, an image with heading, paragraph and button will always have the same markup on any project, I see no point of creating different markup each time project is setup. 

You wanted crazy ideas, here they are! - LOL

**Comment by luboskmetko** [comment]
Nice ideas, guys!

Objects in ITCSS (I propose to switch to it in #1) are such reusable design patterns, check https://github.com/inuitcss?utf8=%E2%9C%93&query=objects

The advantage of objects is that they are, let's say, visually agnostic but can be quickly combined with components to create desired UI

Generic Object
https://github.com/itcss/itcss-netmag/blob/master/css/_objects.media.scss

Project specific Component
https://github.com/itcss/itcss-netmag/blob/master/css/_components.testimonials.scss

HTML snippet:
https://github.com/itcss/itcss-netmag/blob/master/index.html#L76

**Comment by danekszy** [comment]
Yeah, these 2 can be complementary. Reusability is a broad term and therefore we may apply it at various levels of abstractions. Be it patterns, components or just simple mixins or helpers. :)

**Comment by luboskmetko** [comment]
Right :) Just some thoughts on your ideas yet:
- there is a subgenerator task for scaffolding WP projects in the general already - https://github.com/xhtmlized/generator-xh#5-wordpress-development. It's not ideal, just mentioning something is already there :)
- if we should rewrite generator to other tech like Gulp or even just npm, I would rather start a new projects to keep the original generator as it is. It should be easier if we start fresh, especially if we decide to incorporate some more advanced things (eg. replace usemin with scripts bundling, etc).

**Comment by Azzam** [comment]
Hi guys
well i like that ideas so far i got some of points 

<ul>
<li>for drupal i used <a href="https://www.drupal.org/project/static_page" target="_blank">static page module</a></li>
<li>Glup + <a href="https://github.com/postcss/postcss" target="_blank">PostCSS</a> </li>
<li>dropped bootstrap and use <a href="https://www.npmjs.com/package/lost-grid" target="_blank">Lost grid</a></li>
<li>use <a href="https://www.npmjs.com/package/gulp-sass" target="_blank">glup-sass</a> or <a href="https://github.com/jonathantneal/precss"> precss</a></li>
</ul>

also if we create default snippets and components etc we can build our own template for example using something like <a href="https://github.com/Polymer/polymer" target="_blank">Polymer</a> or <a href="https://github.com/x-tag/core" target="_blank">x-tag</a> or <a href="https://customelements.io/" target="_blank">custom elements </a> that will reduce a lot of time too
